### ISSUE TITLE
 routing has a loss of memory >  4 Mb on a problem of 16 nodes

### ISSUE THREAD

**Issue by bardout** [google/or-tools#3294]
**Env**
ortools.version: 9.1.9490
I use the Vehicle Routing Problem Solver, via Python 3.7.8, Windows 10 Pro, archi x64

**Summary**
The routing has a loss of memory, as tracked via psutil, of  4.355 Mb on a routing problem of 16 nodes. Altough python variables are deleted and garbage collected, there seem to be C++ object not deallocated, as _routing, solution and parameters_.
Please point me to the way to deallocate these. No example is provided of memory management via a specific call. 

**Detailled experiments**
I process a large data set, and perform repeated solving routing problem ( between 2 and 12 nodes). it constantly loose on average ~2 Mb per iteration, as tracked via psutil.

it is stable when routing is bypassed. It crashes around 11000 routing calls, exhausting physical mem of 16Gb and virtual memory of 36Gb - from which only ~3Gb is real use in my code, and the loss of memory is from solver.
I am calling gc.collect() every 1k iteration, with cannot get it back. The solving code is based on the tutorial. 

The cost callback  signature is the 2 integer index, and it depends on the data structure, that def should better be **global** instead of local to the problem -  defined once versus repeatedly, but it means data is also global. I have modified code to do this, and it does not impact the memory leak.

There are 3 related issues (2 open, 2 changed in discussion) , which do not help : 
- [3084](https://github.com/google/or-tools/discussions/3084) for CP-SAT with ortools: 9.2.9972, Python 3.8.3, said fixed on master 17/1/22 
- [2967](https://github.com/google/or-tools/discussions/2967) for routing with Python 3.7.2 in 2019, Unanswered
- and [2782](https://github.com/google/or-tools/issues/2782)  on  in C++. 

** Code **
```py
#!/usr/bin/env python3
# [START program]
"""Vehicles Routing Problem (VRP) for bardout #3263. https://github.com/google/or-tools/issues/3263 
 that is geotruth https://forge-is.pf.irt-saintexupery.com/minds/wp2/geotruth/-/issues/107
"""

# [START import]
import time
import math
import numpy as np
import psutil, gc
from ortools.constraint_solver import routing_enums_pb2
from ortools.constraint_solver import pywrapcp
# [END import]

# [START data_model]
def create_data_model():
    """Stores the data for the problem."""
    data = {}
    data['cost'] = np.array( [
        [   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,    0,   0],
        [   0,1000,   6,   6,   6,   7,   7,  10,   7,  10,   9,   7,   9,  10,    9,   7],
        [   0,   6,1000,   6,   8,   5,   5,  10,  10,  10,  10,   9,  10,  10,   10,   5],
        [   0,   6,   6,1000,   6,   6,  10,   6,   6,  10,  10,   6,  10,   8,    6,   6],
        [   0,   6,   8,   6,1000,   9,  10,  10,  10,  10,  10,   6,  10,  10,   10,   6],
        [   0,   7,   5,   6,   9,1000,   7,   7,   7,   7,   5,   7,   9,   5,    5,   7],
        [   0,   7,   5,  10,  10,   7,1000,   8,   7,   7 ,  5,   8,  10,  10,    5,   8],
        [   0,  10,  10,   6,  10,   7,   8,1000,  10,   7,  10,   8,  10,   9,   10,   8],
        [   0,   7,  10,   6,  10,   7,   7,  10,1000,  10,  10,   8,   9,  10,    9,   8],
        [   0,  10,  10,  10,  10,   7,   7,   7,  10,1000,  10,   7,  10,  10,   10,  10],
        [   0,   9,  10 , 10,  10,   5,   5,  10,  10,  10,1000,  10,  10,  10,   10,   9],
        [   0,   7,   9,   6,   6,   7,   8,   8,   8,   7,  10,1000,  10,  10,   10,   8],
        [   0,   9,  10,  10,  10,   9,  10,  10,   9,  10,  10,  10,1000,  10,   10,   9],
        [   0,  10,  10,   8 , 10,   5,  10,   9,  10,  10,  10,  10,  10,1000,   10,  10],
        [   0,   9,  10,   6,  10,   5,   5,  10,   9,  10,  10,  10,  10,  10, 1000,   5],
        [   0,   7,   5,   6,   6,   7,   8,   8,   8,  10,   9,   8,   9,  10,    5,1000],
        ])
    nb_locations = len(data['cost'])
    nb_part = 4
    nb_v_min = 1 # at least one vehicle
    # vehicle number is the max nb of sentence in solution
    nb_v = max( math.ceil(nb_locations / nb_part), nb_v_min)
    print(f'dims: loc:{nb_locations}, part:{nb_part}, v_min:{nb_v_min}, vehicles:{nb_v}') #  16 4 1 1
    data['num_vehicles'] = nb_v
    data['depot'] = 0
    return data
    # [END data_model]


# [START solution_printer]
def print_solution(data, manager, routing, assignment):
    """Prints assignment on console."""
    print(f'Objective: {assignment.ObjectiveValue()}')
    # Display dropped nodes.
    dropped_nodes = 'Dropped nodes:'
    for node in range(routing.Size()):
        if routing.IsStart(node) or routing.IsEnd(node):
            continue
        if assignment.Value(routing.NextVar(node)) == node:
            dropped_nodes += ' {}'.format(manager.IndexToNode(node))
    print(dropped_nodes)
    # Display routes
    total_distance = 0
    for vehicle_id in range(manager.GetNumberOfVehicles()):
        index = routing.Start(vehicle_id)
        plan_output = 'Route for vehicle {}:\n'.format(vehicle_id)
        route_distance = 0
        while not routing.IsEnd(index):
            node_index = manager.IndexToNode(index)
            plan_output += f' {node_index} -> '
            previous_index = index
            index = assignment.Value(routing.NextVar(index))
            route_distance += routing.GetArcCostForVehicle(
                previous_index, index, vehicle_id)
        node_index = manager.IndexToNode(index)
        plan_output += f' {node_index}\n'
        plan_output += 'Distance of the route: {}m\n'.format(route_distance)
        print(plan_output)
        total_distance += route_distance
    print('Total Distance of all routes: {}m'.format(total_distance))
    # [END solution_printer]

@profile 
def main():
    process = psutil.Process() 
    print( 'At begin: ', process.memory_info().rss, gc.get_count() )
    """Solve the CVRP problem."""
    # Instantiate the data problem.
    # [START data]
    data = create_data_model()
    # [END data]

    # Create the routing index manager.
    # [START index_manager]
    manager = pywrapcp.RoutingIndexManager(
            len(data['cost']),
            data['num_vehicles'],
            data['depot'])
    # [END index_manager]

    # Create Routing Model.
    # [START routing_model]
    routing = pywrapcp.RoutingModel(manager)
    # [END routing_model]


    # Create and register a transit callback.
    # [START transit_callback]
    def cost_callback( from_index: int, to_index: int):
        """Returns the COST of arc between the two nodes, or 1000 when unknown."""
        # Convert from routing variable Index to distance matrix NodeIndex.
        i = manager.IndexToNode(from_index)
        j = manager.IndexToNode(to_index)
        return data['cost'][i,j]

    transit_cb = routing.RegisterTransitCallback(cost_callback)
    # [END transit_callback]

    # Define cost of each arc.
    # [START arc_cost]
    routing.SetArcCostEvaluatorOfAllVehicles(transit_cb)
    # [END arc_cost]

    # distance for RVP
    dimension_name = 'Cost'
    routing.AddDimension(
        transit_cb,
        0,  # no slack
        30,  # vehicle maximum travel distance: 5 parts w/ cost 5
        #50,  # vehicle maximum travel distance: 5 parts w/ cost 5
        True,  # start cumul to zero
        dimension_name)
    dimension = routing.GetDimensionOrDie(dimension_name)
    dimension.SetGlobalSpanCostCoefficient(100)

    ## Allow to drop nodes.
    #penalty = 1_000_000
    #for node in range(1, len(data['cost'])):
    #    routing.AddDisjunction([manager.NodeToIndex(node)], penalty)

    # Solve the problem.
    # routing 15 nodes, based on an arc value 1..5, the cost is 10-value in [5.. 10], except 1000 on diagonal
    # search multi routes to cover all nodes, start/end at any node.
    # So the start node 0 has no cost to any node, one column and row if filled with 0
    #Setting first solution heuristic.
    # Setting first solution heuristic.
    # [START parameters]
    params = pywrapcp.DefaultRoutingSearchParameters()
    params.first_solution_strategy = (
            routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC)
    params.local_search_metaheuristic = (
            routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH)
    # Sets a time limit; default is 100 milliseconds.
    params.time_limit.FromSeconds(1)
    #params.lns_time_limit.seconds = 1
    params.solution_limit = 5
    # [END parameters]


    # Solve the problem.
    # [START solve]
    tic = time.perf_counter_ns()
    solution = routing.SolveWithParameters(params)
    elapsed = (time.perf_counter_ns() - tic)//1000
    # [END solve]

    # Print solution on console.
    # [START print_solution]
    if solution:
        print( "one solution", elapsed)
        print_solution(data, manager, routing, solution)
    else:
        print('no solution for relation', elapsed)
    # [END print_solution]

    del data, manager, routing, cost_callback, dimension, transit_cb, params, solution
    gc.collect()
    print( 'At end: ', process.memory_info().rss, gc.get_count() )

if __name__ == '__main__': 
    main() 
# [END program]
```
** Expected **
At the end of problem() , the local variable are deleted, when the interpreter exit the scope,  and memory freed when calling the garbage collector.
 (python variable generally are, but C++ allocation shall be handled in the library wrapped in python API). 

**Observed**
Using the [mprof](https://github.com/pythonprofilers/memory_profiler/),  I reproduce the loss with the code kindly sent by @misuz in [3263](https://github.com/google/or-tools/issues/3263), strictly based on tutorial.
When deleting all local variable and local function _cost_callback_,  that depend on _data_ , -1.379 MiB is reclaimed, but this is a small part of allocated local variables and library objects which are apparently not freed, even after garbage collection. 

At begin:  60788736 (19, 1, 5)
At end:  65155072 (9, 0, 0)  leak:  4689920

After making the callback local, The main contributors are:
- 1.840 MiB           1       routing = pywrapcp.RoutingModel(
-  1.289 MiB           1       params = pywrapcp.DefaultRoutingSearchParameters()
-  2.277 MiB           1       solution = routing.SolveWithParameters(params)
which  total 5.406 MiB , and misc  0.362 but freed   -1.391 MiB, hence  a 4.379 MiB loss 

```
> python -m memory_profiler ref-RVP-mizux.py
At begin:  60788736 (19, 1, 5)
dims: loc:16, part:4, v_min:1, vehicles:4
one solution 410343
At end:  65380352 (30, 0, 0)
Filename: ref-RVP-mizux.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    83   57.973 MiB   57.973 MiB           1   @profile
    84                                         def main():
    85   57.973 MiB    0.000 MiB           1       process = psutil.Process()
    86   57.977 MiB    0.004 MiB           1       print( 'At begin: ', process.memory_info().rss, gc.get_count() )
    87                                             """Solve the CVRP problem."""
    88                                             # Instantiate the data problem.
    89                                             # [START data]
    90   57.977 MiB    0.000 MiB           1       data = create_data_model()
    91                                             # [END data]
    92
    93                                             # Create the routing index manager.
    94                                             # [START index_manager]
    95   57.977 MiB    0.000 MiB           1       manager = pywrapcp.RoutingIndexManager(
    96   57.977 MiB    0.000 MiB           1               len(data['cost']),
    97   57.977 MiB    0.000 MiB           1               data['num_vehicles'],
    98   58.020 MiB    0.043 MiB           1               data['depot'])
    99                                             # [END index_manager]
   100
   101                                             # Create Routing Model.
   102                                             # [START routing_model]
   103   59.832 MiB    1.812 MiB           1       routing = pywrapcp.RoutingModel(manager)
   104                                             # [END routing_model]
   105
   106
   107                                             # Create and register a transit callback.
   108                                             # [START transit_callback]
   109   63.703 MiB    2.234 MiB        3244       def cost_callback( from_index: int, to_index: int):
   110                                                 """Returns the COST of arc between the two nodes, or 1000 when unknown."""
   111                                                 # Convert from routing variable Index to distance matrix NodeIndex.
   112   63.703 MiB    0.004 MiB        3243           i = manager.IndexToNode(from_index)
   113   63.703 MiB    0.000 MiB        3243           j = manager.IndexToNode(to_index)
   114   63.703 MiB    0.000 MiB        3243           return data['cost'][i,j]
   115
   116   59.844 MiB    0.012 MiB           1       transit_cb = routing.RegisterTransitCallback(cost_callback)
   117                                             # [END transit_callback]
   118
   119                                             # Define cost of each arc.
   120                                             # [START arc_cost]
   121   59.852 MiB    0.008 MiB           1       routing.SetArcCostEvaluatorOfAllVehicles(transit_cb)
   122                                             # [END arc_cost]
   123
   124                                             # distance for RVP
   125   59.852 MiB    0.000 MiB           1       dimension_name = 'Cost'
   126   59.852 MiB    0.000 MiB           1       routing.AddDimension(
   127   59.852 MiB    0.000 MiB           1           transit_cb,
   128   59.852 MiB    0.000 MiB           1           0,  # no slack
   129   59.852 MiB    0.000 MiB           1           30,  # vehicle maximum travel distance: 5 parts w/ cost 5
   130                                                 #50,  # vehicle maximum travel distance: 5 parts w/ cost 5
   131   59.852 MiB    0.000 MiB           1           True,  # start cumul to zero
   132   60.020 MiB    0.168 MiB           1           dimension_name)
   133   60.027 MiB    0.008 MiB           1       dimension = routing.GetDimensionOrDie(dimension_name)
   134   60.031 MiB    0.004 MiB           1       dimension.SetGlobalSpanCostCoefficient(100)
   135
   136                                             ## Allow to drop nodes.
   137                                             #penalty = 1_000_000
   138                                             #for node in range(1, len(data['cost'])):
   139                                             #    routing.AddDisjunction([manager.NodeToIndex(node)], penalty)
   140
   141                                             # Solve the problem.
   142                                             # routing 15 nodes, based on an arc value 1..5, the cost is 10-value in [5.. 10], except 1000 on diagonal
   143                                             # search multi routes to cover all nodes, start/end at any node.
   144                                             # So the start node 0 has no cost to any node, one column and row if filled with 0
   145                                             #Setting first solution heuristic.
   146                                             # Setting first solution heuristic.
   147                                             # [START parameters]
   148   61.453 MiB    1.422 MiB           1       params = pywrapcp.DefaultRoutingSearchParameters()
   149                                             params.first_solution_strategy = (
   150   61.465 MiB    0.012 MiB           1               routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC)
   151                                             params.local_search_metaheuristic = (
   152   61.465 MiB    0.000 MiB           1               routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH)
   153                                             # Sets a time limit; default is 100 milliseconds.
   154   61.465 MiB    0.000 MiB           1       params.time_limit.FromSeconds(1)
   155                                             #params.lns_time_limit.seconds = 1
   156   61.465 MiB    0.000 MiB           1       params.solution_limit = 5
   157                                             # [END parameters]
   158
   159
   160                                             # Solve the problem.
   161                                             # [START solve]
   162   61.465 MiB    0.000 MiB           1       tic = time.perf_counter_ns()
   163   63.715 MiB    0.012 MiB           1       solution = routing.SolveWithParameters(params)
   164   63.715 MiB    0.000 MiB           1       elapsed = (time.perf_counter_ns() - tic)//1000
   165                                             # [END solve]
   166
   167                                             # Print solution on console.
   168                                             # [START print_solution]
   169   63.715 MiB    0.000 MiB           1       if solution:
   170   63.715 MiB    0.000 MiB           1           print( "one solution", elapsed)
   171   63.742 MiB    0.027 MiB           1           print_solution(data, manager, routing, solution)
   172                                             else:
   173                                                 print('no solution for relation', elapsed)
   174                                             # [END print_solution]
   175
   176   62.352 MiB   -1.391 MiB           1       del data, manager, routing, dimension, transit_cb, params, solution
   177   62.352 MiB    0.000 MiB           1       gc.collect()
   178   62.352 MiB    0.000 MiB           1       print( 'At end: ', process.memory_info().rss, gc.get_count() )


```

**Comment by lperron** [comment]
I did not manage to run the code with the @profile tag

Anyway, without it I get:

```
At begin:  32452608 (92, 1, 6)
dims: loc:16, part:4, v_min:1, vehicles:4
one solution 11475
Objective: 2468
Dropped nodes:
Route for vehicle 0:
 0 ->  12 ->  15 ->  7 ->  9 ->  0
Distance of the route: 24m

Route for vehicle 1:
 0 ->  2 ->  1 ->  0
Distance of the route: 6m

Route for vehicle 2:
 0 ->  8 ->  3 ->  11 ->  4 ->  0
Distance of the route: 18m

Route for vehicle 3:
 0 ->  14 ->  6 ->  10 ->  5 ->  13 ->  0
Distance of the route: 20m

Total Distance of all routes: 68m
At end:  37326848 (16, 0, 0)
```

**Comment by lperron** [comment]
This is not memory loss. This is C++ memory not reclaimed by the process.

Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by lperron** [comment]
I checked all C++ dtor are called.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by bardout** [comment]
Knowing it is not the C+lib, I assume it has to be python behavior on this tutorial code, or the wrapper. In my code, I already explicitely **del** statement, and  also tried using **slots** in a class for all python variables. In other python large code I wrote, handling many objects, Ithe OS seems to recelaim memory given back by python. 

The  other option is **inside the wrapper**. As I am unable to debug the python wrapper on the 3 functions that allocate memory, I have no workaround to use the lib on a large set.  

I checked that issue on _linux "Ubuntu 20.04.3 LTS"_ with worse result than _Windows 10_: 
```
At begin:  86380544 (133, 0, 5)
dims: loc:16, part:4, v_min:1, vehicles:4
one solution 25411
At end:  93691904 (18, 0, 0)  leak:  7311360
```

Further ideas: from old thread [2011](https://web.archive.org/web/20110621221104/http://effbot.org/pyfaq/why-doesnt-python-release-the-memory-when-i-delete-a-large-object.htm ) discuss for python 2.x, and [2018](https://stackoverflow.com/questions/51938963/python-memory-not-being-released-on-linux) is quite detailed about that. I tried on linux to call  _malloc_trim()_ as suggested, to give back the collected memory by python to C heap, with effect to grow the rss  by 0.488 MiB instead of reducing.
